### DISCUSSION TITLE
Contribution ?

### DISCUSSION THREAD

**Discussion by TOomaAh** [henrygd/beszel#95] [Support]
Hello,

I use your project and I think I can contribute to the code if you wish.
Is it possible to contribute to the file structure or even to the code by submitting PRs?

**Comment by henrygd** [comment]
Thanks for your interest! I'm definitely open to contributions.

However, the project is still very new and I'm making updates frequently, so it would be nice to discuss beforehand what contributions you have in mind.

**Comment by TOomaAh** [comment]
Of course! 
For example, the folder structure doesn't allow you to share models easily between the agent and the server. For example, the SystemInfo, SystemData structures or the Version variable are declared in both the agent and the server.

I use this structure a lot for my projects, for example:
https://github.com/evrone/go-clean-template

Without even following this template, perhaps create a "beszel" parent module where you can create the common entities and have 2 sub-modules, one for the agent and one for the server, containing the main function for each.

**Reply by henrygd** [reply]
That would be very helpful, but we would need to make sure that we can adjust the existing GitHub workflows to build separate binaries and docker images for the hub and agent as it does now.

I had some trouble setting that up initially which is why I ended up separating everything. I only started learning Go this year though, so I'm not well versed in different project structures.

If you know a better way to do it that maintains the workflow functionality, then I'm definitely open to restructuring.

**Reply by TOomaAh** [reply]
Here's a first draft of what it might look like. I think it's far from perfect and there's still a lot of work to be done. I'm not a golang professional, but the agent and the hub cohabit in the same project like this.

If you like, I can create a draft pull request so you can comment on what works for you and what doesn't. That way I can get your feedback.

https://github.com/TOomaAh/beszel/tree/feature/change-file-structure

**Reply by henrygd** [reply]
Thanks, you can make a pull request and I'll check it out this weekend.
